### PR TITLE
[chore] Method that sorts Datapoint Slice by Attribute key

### PR DESCRIPTION
**Pull request by BominRahmani**
Adds a method that allows Metrics Datapoint slices of Golden files to be sorted by their respective attribute key. This is in the efforts to bring Golden files 1 step closer to being more deterministic.

There is a test folder (sort-datapoint-slice) in golden/testdata that shows an example of what this accomplishes.

### CODE REVIEW

File: internal/coreinternal/golden/sort_metrics.go [hunk 1]
```diff
@@ -85,3 +87,75 @@ func sortAttributeMap(mp pcommon.Map) {
...
+		return true
+	})
+	b.Range(func(k string, _ pcommon.Value) bool {
+		bKeys = append(bKeys, k)
+		return true
+	})
+
+	for i := 0; i < len(aKeys); i++ {
+		if aKeys[i] != bKeys[i] {
+			return strings.Compare(aKeys[i], bKeys[i])
```
Comment (djaglowski) [inline internal/coreinternal/golden/sort_metrics.go:156]: Don't we need to sort `aKeys` and `bKeys` before comparing elements?

Comment (BominRahmani) [reply internal/coreinternal/golden/sort_metrics.go:156]: This takes advantage of the fact that the attribute keys should have been sorted previously by the work done in the previous MR where we sort the attributes. The sort datapoint slice is called after that main sort where all the attributes are sorted so it should ensure these are always presorted.

File: internal/coreinternal/golden/sort_metrics.go [hunk 2]
```diff
@@ -5,12 +5,13 @@ package golden // import "github.com/open-telemetry/opentelemetry-collector-cont
...
 import (
 	"sort"
+	"strings"
 
 	"go.opentelemetry.io/collector/pdata/pcommon"
 	"go.opentelemetry.io/collector/pdata/pmetric"
 )
 
-// sorts all Resource Metrics attributes and Datapoint Slice metric attributes
+// sorts all Resource Metrics attributes and Datapoint Slice metric attributes and all Datapoint slices
```
Comment (djaglowski) [inline internal/coreinternal/golden/sort_metrics.go:14]:
```suggestion
// sorts all attribute maps in a pmetric.Metrics
```

File: internal/coreinternal/golden/sort_metrics.go [hunk 3]
```diff
@@ -51,6 +52,7 @@ func sortMetrics(ms pmetric.Metrics) {
 			}
 		}
 	}
+	sortMetricDataPointSlices(ms)
```
Comment (djaglowski) [inline internal/coreinternal/golden/sort_metrics.go:55]:
```suggestion
```
Let's also rename `sortMetrics` to `sortMetricAttributes`. This way, we can call one or both of the functions in either order. (This becomes important later if we turn these into options.)

File: internal/coreinternal/golden/sort_metrics.go [hunk 4]
```diff
@@ -85,3 +87,77 @@ func sortAttributeMap(mp pcommon.Map) {
 	}
 	tempMap.CopyTo(mp)
 }
+
+// sortMetricDataPointSlices sorts the datapoint slice of a pmetric.Metrics according to the alphanumeric ordering of map key
```
Comment (djaglowski) [inline internal/coreinternal/golden/sort_metrics.go:91]:
```suggestion
// sortMetricDataPointSlices sorts the datapoint slice according to the alphanumeric ordering of attribute keys
```

File: receiver/mysqlreceiver/testdata/scraper/expected.yaml [hunk 5]
```diff
@@ -1666,7 +1666,7 @@ resourceMetrics:
                   startTimeUnixNano: "1644862687825728000"
                   timeUnixNano: "1644862687825772000"
               isMonotonic: true
-            unit: "s"
+            unit: s
```
Comment (djaglowski) [inline receiver/mysqlreceiver/testdata/scraper/expected.yaml:1669]: This appears unrelated

File: internal/coreinternal/golden/testdata/sort-datapoint-slice/before.yaml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+                    - key: b
+                      value:
+                        stringValue: BBBBFirst
+                  startTimeUnixNano: "1111111111111111111"
+                  timeUnixNano: "1111111111111111111"
+                - asInt: "0"
+                  attributes:
+                    - key: b
+                      value:
+                        stringValue: BBBBSecond
```
Comment (djaglowski) [inline internal/coreinternal/golden/testdata/sort-datapoint-slice/before.yaml:45]: This PR doesn't incorporate attribute values, right?

File: internal/coreinternal/golden/testdata/sort-datapoint-slice/before.yaml [hunk 7]
```diff
@@ -0,0 +1,52 @@
+resourceMetrics:
+  - resource:
+    scopeMetrics:
+      - metrics:
+          - description: The number of bytes which have flowed through the network interface.
+            name: nsxt.node.network.io
+            sum:
+              aggregationTemporality: 2
+              dataPoints:
```
Comment (djaglowski) [inline internal/coreinternal/golden/testdata/sort-datapoint-slice/before.yaml:9]: Can we have some additional data points that exercise the sorting of attributes _within_ a map? Further, can we have some that require comparison beyond the first key?

- a
- b, a
- c, b
- c, b, a
- c, a, d

File: internal/coreinternal/golden/testdata/sort-datapoint-slice/before.yaml [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+                - asInt: "0"
+                  attributes:
+                    - key: a
+                      value:
+                        stringValue: AAAA
+                  startTimeUnixNano: "1111111111111111111"
+                  timeUnixNano: "1111111111111111111"
```
Comment (djaglowski) [inline internal/coreinternal/golden/testdata/sort-datapoint-slice/before.yaml:16]: You should be able to omit these fields for this test, as they only distract from the point of the test.

File: internal/coreinternal/golden/sort_metrics_test.go [hunk 9]
```diff
@@ -42,3 +44,15 @@ func TestSortAttributes(t *testing.T) {
...
+func TestSortMetricDatapointSlice(t *testing.T) {
+	beforePath := filepath.Join("testdata", "sort-datapoint-slice", "before.yaml")
+	afterPath := filepath.Join("testdata", "sort-datapoint-slice", "after.yaml")
+	before, err := ReadMetrics(beforePath)
+	require.NoError(t, err)
+	after, err := ReadMetrics(afterPath)
+	require.NoError(t, err)
+	sortMetricDataPointSlices(before)
+	require.Equal(t, before, after)
+
```
Comment (djaglowski) [inline internal/coreinternal/golden/sort_metrics_test.go:57]:
```suggestion
```

**Review by djaglowski — COMMENTED** [review]
Please also regenerate all golden files so we can see what the expected change is.

File: internal/coreinternal/golden/testdata/sort-datapoint-slice/before.yaml (hunk 6 above)
Comment (BominRahmani) [reply internal/coreinternal/golden/testdata/sort-datapoint-slice/before.yaml:45]: It does not, I just left that there so the tester can see how it interacts when the sorting of the slices encounters two of the same keys. In essence showing the first one it encounters gets priority in the sort.

File: receiver/mysqlreceiver/testdata/scraper/expected.yaml (hunk 5 above)
Comment (BominRahmani) [reply receiver/mysqlreceiver/testdata/scraper/expected.yaml:1669]: This is the only change from running golden test on all the receivers.

**Comment by BominRahmani** [comment]
> Please also regenerate all golden files so we can see what the expected change is.

This is already all the changes when running the sort datapoint slice.

File: receiver/mysqlreceiver/testdata/scraper/expected.yaml (hunk 5 above)
Comment (djaglowski) [reply receiver/mysqlreceiver/testdata/scraper/expected.yaml:1669]: I think it's unrelated to these changes but it's trivial so we'll just pick it up.

**Review by djaglowski — COMMENTED** [review]
> This is already all the changes when running the sort datapoint slice.

Ok, I suppose this makes sense because we generally want data points to have the same attribute keys. 

Since this PR doesn't improve the golden files, I think it makes sense to extend it to cover resources and scopes, where I expect we'll see meaningful improvements.


For data points, I think a followup PR will make sense, which layers on sorting by value, since the values tend to be static or even enumerated.

**Review by djaglowski — APPROVED** [review]

**Comment by djaglowski** [comment]
Thanks @BominRahmani, @dehaansa!
